packages/common/src/errors.ts

```diff
@@ -2,43 +2,74 @@ import { StandardizedFilePath } from "./fileSystem";
 import { getSyntaxKindName } from "./helpers";
 import { ts } from "./typescript";
 
+/** Minimal attributes to show a error message with source */
+export type MaybeTraceAbleNode = undefined | null | {} | {
+    // TODO: Find correct type
+    getSourceFile: () => {
+        getFilePath: () => StandardizedFilePath;
+        getText: () => string;
+    }
+    getPos(): number;
+}
+
 /** Collection of helper functions that can be used to throw errors. */
 export namespace errors {
   /** Base error class. */
   export abstract class BaseError extends Error {
+    public readonly source: { fileName: string; pos: ts.LineAndCharacter } | undefined;
     /** @private */
-    constructor(public readonly message: string) {
-      super(message);
-
-      this.message = message;
+    constructor(public readonly message: string, node?: MaybeTraceAbleNode) {
+      let messageWithSource = message;
+      let source: { fileName: string; pos: ts.LineAndCharacter } | undefined;
+      if (node && "getSourceFile" in node && "getPos" in node) {
+        try {
+          const sourceFile = node.getSourceFile();
+          const sourceCode = sourceFile.getText();
+          const pos = node.getPos();
+          const textBeforePos = sourceCode.substring(0, pos);
+          const line = textBeforePos.match(/\n/g)?.length || 0;
+          const brokenLineStart = textBeforePos.lastIndexOf("\n", pos);
+          const brokenLineEnd = sourceCode.indexOf("\n", pos);
+          const brokenLine = sourceCode.substring(brokenLineStart + 1, brokenLineEnd === -1 ? undefined : brokenLineEnd);
+          source = { fileName: sourceFile.getFilePath(), pos: { line, character: pos - brokenLineStart } };
+          const linePrefix = `> ${source.pos.line + 1} |`;
+          messageWithSource += `\nin ${source.fileName}:${source.pos.line + 1}:${source.pos.character + 1}\n\n${linePrefix}${brokenLine}\n${" ".repeat(linePrefix.length - 1)}|${" ".repeat(source.pos.character)}^`;
+        } catch (e) {
+          // Errors inside BaseError would be confusing
+          // so ignore errors here and fallback to the original message
+        }
+    }
```

Comment:
Would be good if we added some tests for this.

Reply:
Added some tests :)  
I am not sure how to setup the test inside `packages/common` so I had to place the test inside `packages/ts-morph`